### PR TITLE
INTYGFV-15653: Added ErrorPage to handle error routes.

### PR DESCRIPTION
**Pull request by ChrilleSigerhed**

### CODE REVIEW

File: apps/rehabstod/src/components/Error/ErrorCode.ts
Comment (SamuelTilly) [inline apps/rehabstod/src/components/Error/ErrorCode.ts:1]: This should be added to our `errorSchema` so that it can be used by `errorCode` there as well.

```typescript
export enum ErrorCode {
  LOGIN_FAILED = 'LOGIN_FAILED',
  LOGIN_HSA_ERROR = 'LOGIN_HSA_ERROR',
  LOGIN_MEDARBETARUPPDRAG_SAKNAS = 'LOGIN_MEDARBETARUPPDRAG_SAKNAS',
  LOGIN_SAKNAR_HSA_REHABROLL = 'LOGIN_SAKNAR_HSA_REHABROLL',
}

export const ErrorCodeEnum = z.nativeEnum(ErrorCode)

export const errorSchema = z.object({
  errorId: z.string(),
  errorCode: ErrorCodeEnum,
  message: z.string(),
  stackTrace: nullable(z.string()),
})
```

File: apps/rehabstod/src/store/errorMiddleware.ts [hunk 1]
```diff
@@ -13,17 +13,21 @@ export const errorMiddleware: Middleware =
...
+      const errorMessage = `${message}' method '${method}' url '${url}`
       const errorId = uuidv4()
       const errorData = {
         errorId,
         errorCode,
         message: errorMessage,
         stackTrace: null,
       }
-      dispatch(api.endpoints.logError.initiate({ errorData, ...errorData }))
+      dispatch(api.endpoints.logError.initiate({ ...errorData, errorData }))
```
Comment (SamuelTilly) [inline apps/rehabstod/src/store/errorMiddleware.ts:30]: This should be `{ errorData }` without the extra spread. Not sure why typescript is not complaining here.

**Review by SamuelTilly — CHANGES_REQUESTED** [review]
Looks good so far. I'll test it a bit closer, but until then here are some things that can be improved

File: apps/rehabstod/src/components/Error/ErrorCode.ts
Comment (ChrilleSigerhed) [reply apps/rehabstod/src/components/Error/ErrorCode.ts:1]: Moved texts & titles into this schema aswell.

File: apps/rehabstod/src/pages/Error/ErrorPage.tsx [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  const dispatch = useAppDispatch()
+  const [displayCopyMessage, setDisplayCopyMessage] = useState(false)
+  const { current: errorId } = useRef(uuidv4())
+  const { routingErrorId } = useAppSelector((state) => state.error)
+  const { errorCode } = useAppSelector((state) => state.error)
+
+  useEffect(() => {
+    if (location.search) {
+      const params = new URLSearchParams(location.search)
+      const reason = params.get('reason') ?? ''
```
[the remark this answers is not in the document]
Comment (SamuelTilly) [reply apps/rehabstod/src/pages/Error/ErrorPage.tsx:33]: This would not be needed with the changes suggested in ErrorPage component with react-router usage.

[the remark this answers is not in the document]
Comment (ChrilleSigerhed) [reply apps/rehabstod/src/pages/Error/ErrorPage.tsx:33]: I had some troubles with getting the provided solution with react-router usage to work, so I decided to go with useSearchParams for now. If you would prefer we use routes instead, I will make an effort to make it work!

File: apps/rehabstod/src/pages/Error/ErrorPage.tsx [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        api.endpoints.logError.initiate({
+          errorData: {
+            errorId,
+            errorCode: generatedErrorCode,
+            message: reason,
+            stackTrace: null,
+          },
+        })
+      )
+      dispatch(setErrorCode(generatedErrorCode))
```
[the remark this answers is not in the document]
Comment (ChrilleSigerhed) [reply apps/rehabstod/src/pages/Error/ErrorPage.tsx:45]: It was previously, but not needed anymore! Thanks for noticing!

File: apps/rehabstod/src/pages/Error/ErrorPage.test.tsx [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import { ErrorPage } from './ErrorPage'
+import { setErrorCode } from '../../store/slices/error.slice'
+import { store } from '../../store/store'
+import { ErrorCode, ErrorTitle } from '../../components/Error/ErrorCode'
+
+const renderComponent = () => {
+  renderWithRouter(<ErrorPage />)
+}
+
+describe('ErrorPage tests', () => {
```
[the remark this answers is not in the document]
Comment (ChrilleSigerhed) [reply apps/rehabstod/src/pages/Error/ErrorPage.test.tsx:13]: Done!

File: apps/rehabstod/src/pages/Error/ErrorPage.tsx
[the remark this answers is not in the document]
Comment (ChrilleSigerhed) [reply apps/rehabstod/src/pages/Error/ErrorPage.tsx:1]: Done!

File: apps/rehabstod/src/pages/Error/ErrorPage.tsx [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    setDisplayCopyMessage(true)
+    navigator.clipboard.writeText(routingErrorId)
+  }
+  return (
+    <IDSContainer>
+      <PageHero type="error" icon="attention">
+        <DisplayRoutingError errorCode={errorCode} />
+        <div className="flex justify-center">
+          <p className="mr-2 font-bold">FEL-ID:</p>
+          {routingErrorId}
```
[the remark this answers is not in the document]
Comment (ChrilleSigerhed) [reply apps/rehabstod/src/pages/Error/ErrorPage.tsx:59]: Done!

**Comment by ChrilleSigerhed** [comment]
@SamuelTilly I had some problems with following the route example, using errorContext & hook. If you feel like this is the solution, maybe we can sync up and figure it out together?

File: apps/rehabstod/src/schemas/errorSchema.ts [hunk 6]
```diff
@@ -9,9 +9,26 @@ export enum ErrorCode {
...
+  LOGIN_MEDARBETARUPPDRAG_SAKNAS = 'Medarbetaruppdrag saknas',
+  LOGIN_SAKNAR_HSA_REHABROLL = 'Behörighet saknas',
+}
+
+export enum ErrorText {
+  LOGIN_FAILED = 'Gå tillbaka till startsidan och försök igen. Om felet kvarstår, kontakta i första hand din lokala IT-support och i andra hand ',
+  LOGIN_HSA_ERROR = 'Försök igen om en stund. Om felet kvarstår, kontakta i första hand din lokala IT-support och i andra hand ',
+  LOGIN_MEDARBETARUPPDRAG_SAKNAS = "Det krävs minst ett giltigt medarbetaruppdrag med ändamål 'Vård och behandling' för att använda Rehabstöd.",
+  LOGIN_SAKNAR_HSA_REHABROLL = 'För att logga in som Rehabkoordinator krävs att du har den rollen för vårdenheten i HSA. Kontakta din lokala HSA-administratör för behörighet.',
 }
```
Comment (SamuelTilly) [inline apps/rehabstod/src/schemas/errorSchema.ts:27]: Let's move these to the components where they are used as we discussed 😄

File: apps/rehabstod/src/store/slices/error.slice.ts
Comment (SamuelTilly) [inline apps/rehabstod/src/store/slices/error.slice.ts:1]: error slice is no longer used for error page, so we can probably revert these changes

Comment (ChrilleSigerhed) [reply apps/rehabstod/src/store/slices/error.slice.ts:1]: Done!

File: apps/rehabstod/src/schemas/errorSchema.ts (hunk 6 above)
Comment (ChrilleSigerhed) [reply apps/rehabstod/src/schemas/errorSchema.ts:27]: Done!

**Review by SamuelTilly — APPROVED** [review]
Awesome work! 🍾
